libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -0,0 +1,158 @@
+
+    // @Param: BW
+    // @DisplayName: Bandwidth
+    // @Description: Dynamic notch bandwidth in Hz
+    // @Range: 5 100
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("BW", 3, HarmonicNotchFilterParams, _bandwidth_hz, 20),
+
+    // @Param: ATT
+    // @DisplayName: Attenuation
+    // @Description: Dynamic notch attenuation in dB
+    // @Range: 5 30
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
+    // @DisplayName: Harmonics
+    // @Description: Number of harmonic frequencies to add to the filter. This option takes effect on the next reboot.
+    // @Range: 0 2
+    // @User: Advanced
+    AP_GROUPINFO("HMNCS", 5, HarmonicNotchFilterParams, _harmonics, 1),
+
+    AP_GROUPEND
+};
+
+
+/*
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::init(float _sample_freq_hz, float _center_freq_hz, float _bandwidth_hz, float _attenuation_dB)
+{
+    if (filters == nullptr) {
+        return;
+    }
+
+    sample_freq_hz = _sample_freq_hz;
+    bandwidth_hz = _bandwidth_hz;
+    attenuation_dB = _attenuation_dB;
+
+    // adjust the center frequency to be in the allowable range
+    _center_freq_hz = constrain_float(_center_freq_hz, _bandwidth_hz / 1.98f, sample_freq_hz * 0.48f);
+
+    float octaves = log2f(_center_freq_hz  / (_center_freq_hz - bandwidth_hz/2.0f)) * 2.0f;
+    float A = powf(10, -attenuation_dB/40.0f);
+    float Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
+
+    for (int i=0; i<harmonics+1; i++) {
+        filters[i]._init(sample_freq_hz, _center_freq_hz * (i+1), A, Q);
+    }
+    initialised = true;
+}
+
+/*
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::create(uint8_t _harmonics)
+{
+    filters = new NotchFilter<T>[_harmonics+1];
+    harmonics = _harmonics;
+}
+
+template <class T>
+void HarmonicNotchFilter<T>::update(float center_freq_hz)
+{
+    if (!initialised) {
+        return;
+    }
+
+    // adjust the center frequency to be in the allowable range
+    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz / 1.98f, sample_freq_hz * 0.48f);
+
+    float octaves = log2f(center_freq_hz  / (center_freq_hz - bandwidth_hz/2.0f)) * 2.0f;
+    float A = powf(10, -attenuation_dB/40.0f);
+    float Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
+
+    for (int i=0; i<harmonics+1; i++) {
+        filters[i]._init(sample_freq_hz, center_freq_hz * (i+1), A, Q);
+    }
+}
+
+template <class T>
+T HarmonicNotchFilter<T>::apply(const T &sample)
+{
+    if (!initialised) {
+        return sample;
+    }
+
+    T output = filters[0].apply(sample);
+    for (int i=1; i<harmonics+1; i++) {
+        output = filters[i].apply(output);
+    }
+    return output;
+}
+
+template <class T>
+void HarmonicNotchFilter<T>::reset()
+{
+    if (!initialised) {
+        return;
+    }
+
+    for (int i=0; i<harmonics+1; i++) {
```

Comment:
small thing but we don't normally use "int".. more often "int16_t" or "int8_t"

Reply:
Thanks, I'll clean this up with the variable naming Leonard wanted

Reply:
Ok fixed